### PR TITLE
Custom Brotli Dictionaries

### PR DESCRIPTION
**Pull request by rachel-bousfield**
Adds support for custom brotli dictionaries, allowing us to dramatically cut Stylus contract sizes.

Resolves STY-67
- https://github.com/OffchainLabs/stylus-geth/pull/34

### CODE REVIEW

File: arbitrator/brotli/src/cgo.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+use core::{mem::MaybeUninit, slice};
+
+/// Mechanism for passing data between Go and Rust where Rust can specify the initialized length.
+#[derive(Clone, Copy)]
+#[repr(C)]
+pub struct BrotliBuffer {
+    /// Points to data owned by Go.
+    ptr: *mut u8,
+    /// The length in bytes.
+    len: *mut usize,
```
Comment (mikebenfield) [inline arbitrator/brotli/src/cgo.rs:14]: It's not clear to me why the len is also a pointer - couldn't it just be a usize?

File: arbitrator/brotli/src/dicts/mod.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        match self {
+            Self::StylusProgram => Some(include_bytes!("stylus-program-11.lz")),
+            _ => None,
+        }
+    }
+
+    /// Returns a pointer to a compression-ready instance of the given dictionary.
+    pub fn ptr(&self, level: u32) -> Option<*const EncoderPreparedDictionary> {
+        match self {
+            Self::StylusProgram if level == 11 => Some(STYLUS_PROGRAM_DICT.0),
```
Comment (mikebenfield) [inline arbitrator/brotli/src/dicts/mod.rs:78]: Why the check `if level == 11`? brotli can still use a custom dictionary at lower levels can't it?

**Review by mikebenfield — COMMENTED** [review]
Just a couple small remarks.

File: arbitrator/brotli/src/dicts/mod.rs (hunk 2 above)
Comment (rachel-bousfield) [reply arbitrator/brotli/src/dicts/mod.rs:78]: It seems you need a different dictionary for each level. Ours is a level 11 and that's what `cargo stylus` uses

File: arbitrator/brotli/src/cgo.rs (hunk 1 above)
Comment (rachel-bousfield) [reply arbitrator/brotli/src/cgo.rs:14]: Added another comment :)

File: arbitrator/brotli/src/dicts/mod.rs (hunk 2 above)
Comment (rachel-bousfield) [reply arbitrator/brotli/src/dicts/mod.rs:78]: You know, we should probs make this return a result given that. I'll add a commit

Comment (mikebenfield) [reply arbitrator/brotli/src/dicts/mod.rs:78]: Ah, I see what you mean. The binary dictionary we supply isn't level-specific, but the thing prepared by `BrotliEncoderPrepareDictionary` is. In that case it seems like maybe we'd want to either allow the possibility to prepare dictionaries at other levels, or else only allow compression at level 11. But I don't have a strong opinion; as-is is fine with me.

File: arbitrator/brotli/src/cgo.rs (hunk 1 above)
Comment (mikebenfield) [reply arbitrator/brotli/src/cgo.rs:14]: Ah right; I somehow missed that that actually happened. The comment probably isn't necessary; should be implied by the type; I just missed it.

**Review by mikebenfield — APPROVED** [review]
LGTM; I did leave another couple remarks.

**Pull request merged**